examples/poisson_solver/poisson_solver.cpp

```diff
@@ -0,0 +1,240 @@
+    }
+}
+
+
+// Generates the RHS vector given `f` and the boundary conditions.
+template <typename Closure>
+void generate_rhs(Closure f, double u0, double u1, gko::matrix::Dense<> *rhs)
+{
+    const auto discretization_points = rhs->get_size().num_rows;
+    auto values = rhs->get_values();
+    const auto h = 1.0 / (discretization_points + 1);
+    for (int i = 0; i < discretization_points; ++i) {
+        const auto xi = (i + 1) * h;
+        values[i] = -f(xi) * h * h;
+    }
+    values[0] += u0;
+    values[discretization_points - 1] += u1;
+}
+
+
+// Prints the solution `u`.
+void print_solution(double u0, double u1, const gko::matrix::Dense<> *u)
+{
+    std::cout << u0 << '\n';
+    for (int i = 0; i < u->get_size().num_rows; ++i) {
+        std::cout << u->get_const_values()[i] << '\n';
+    }
+    std::cout << u1 << std::endl;
+}
+
+
+// Computes the 1-norm of the error given the computed `u` and the correct
+// solution function `correct_u`.
+template <typename Closure>
+double calculate_error(int discretization_points, const gko::matrix::Dense<> *u,
+                       Closure correct_u)
+{
+    const auto h = 1.0 / (discretization_points + 1);
+    auto error = 0.0;
+    for (int i = 0; i < discretization_points; ++i) {
+        using std::abs;
+        const auto xi = (i + 1) * h;
+        error +=
+            abs(u->get_const_values()[i] - correct_u(xi)) / abs(correct_u(xi));
+    }
+    return error;
+}
+
+
+int main(int argc, char *argv[])
+{
+    // Some shortcuts
+    using vec = gko::matrix::Dense<double>;
+    using mtx = gko::matrix::Csr<double, int>;
+    using cg = gko::solver::Cg<double>;
+    using bj = gko::preconditioner::BlockJacobiFactory<>;
+
+    if (argc < 2) {
+        std::cerr << "Usage: " << argv[0] << " DISCRETIZATION_POINTS [executor]"
+                  << std::endl;
+        std::exit(-1);
+    }
+
+    // Get number of discretization points
+    const int discretization_points = argc >= 2 ? std::atoi(argv[1]) : 100;
+    const auto executor_string = argc >= 3 ? argv[2] : "reference";
+
+    // Figure out where to run the code
+    const auto omp = gko::OmpExecutor::create();
+    std::map<std::string, std::shared_ptr<gko::Executor>> exec_map{
+        {"omp", omp},
+        {"cuda", gko::CudaExecutor::create(0, omp)},
+        {"reference", gko::ReferenceExecutor::create()}};
+
+    // executor where Ginkgo will perform the computation
+    const auto exec = exec_map.at(executor_string);  // throws if not valid
+    // executor used by the application
+    const auto app_exec = exec_map["omp"];
+
+    // problem:
+    auto correct_u = [](double x) { return x * x * x; };
+    auto f = [](double x) { return 6 * x; };
+    auto u0 = correct_u(0);
+    auto u1 = correct_u(1);
+
+    // initialize matrix and vectors
+    auto matrix = mtx::create(app_exec, gko::dim(discretization_points),
+                              3 * discretization_points - 2);
+    generate_stencil_matrix(lend(matrix));
```

Comment:
Maybe a comment on the usage of  `lend` to make this example self-explanatory ? Because due to the the `auto` type of `matrix`, the actual type is hidden from the user. So, it may be a bit confusing when to use `lend`.

Reply:
not sure how to explain it properly without having a long discussion on it. I think for that we would need to write the tutorials, and point people to it.